chalk-solve/src/clauses.rs

```diff
@@ -952,3 +953,109 @@ impl ToProgramClauses for AssociatedTyDatum {
         );
     }
 }
+
+/// Returns a set of program clauses that could possibly match
+/// `goal`. This can be any superset of the correct set, but the
+/// more precise you can make it, the more efficient solving will
+/// be.
+pub fn program_clauses_that_could_match(
+    program: &dyn RustIrDatabase,
+    goal: &DomainGoal,
+    vec: &mut Vec<ProgramClause>,
+) {
+    let mut clauses = vec![];
```

Comment:
If we adopt the "outer, public" function I mentioned above, we can remove this intermediate vector and just write directly to `vec`